http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutesJsonCodec.scala

```diff
@@ -0,0 +1,295 @@
+        workerMachineType <- c.downField("workerMachineType").as[Option[String]]
+        workerDiskSize <- c.downField("workerDiskSize").as[Option[Int]].flatMap(x => if(x.exists(_ < 0)) Left(negativeNumberDecodingFailure) else Right(x))
+        numberOfWorkerLocalSSDs <- c.downField("numberOfWorkerLocalSSDs").as[Option[Int]].flatMap(x => if(x.exists(_ < 0)) Left(negativeNumberDecodingFailure) else Right(x))
+        numberOfPreemptibleWorkers <- c.downField("numberOfPreemptibleWorkers").as[Option[Int]].flatMap(x => if(x.exists(_ < 0)) Left(negativeNumberDecodingFailure) else Right(x))
+        res <- numberOfWorkersInput match {
+          case Some(x) if x < 0 => Left(negativeNumberDecodingFailure)
+          case Some(0) => Right(RuntimeConfigRequest.DataprocConfig(Some(0), masterMachineType, masterDiskSize))
+          case Some(1) => Left(oneWorkerSpecifiedDecodingFailure)
+          case Some(x) => Right(RuntimeConfigRequest.DataprocConfig(Some(x), masterMachineType, masterDiskSize, workerMachineType, workerDiskSize, numberOfWorkerLocalSSDs, numberOfPreemptibleWorkers))
+          case None =>
+            Right(RuntimeConfigRequest.DataprocConfig(None, masterMachineType, masterDiskSize, workerMachineType, workerDiskSize, numberOfWorkerLocalSSDs, numberOfPreemptibleWorkers))
+        }
+      } yield res
+  }
+
+  implicit val gceConfigDecoder: Decoder[model.RuntimeConfigRequest.GceConfig] = Decoder.forProduct2(
+    "machineType",
+    "diskSize"
+  )((mt, ds) => RuntimeConfigRequest.GceConfig(mt, ds))
+
+  implicit val runtimeConfigDecoder: Decoder[model.RuntimeConfigRequest] = Decoder.instance {
+    x =>
+      //For newer version of requests, we use `cloudService` field to distinguish whether user is
+      val newDecoder = for {
+        cloudService <- x.downField("cloudService").as[CloudService]
+        r <- cloudService match {
+          case CloudService.Dataproc =>
+            x.as[RuntimeConfigRequest.DataprocConfig]
+          case CloudService.GCE =>
+            x.as[RuntimeConfigRequest.GceConfig]
+        }
+      } yield r
+
+      newDecoder // when the request has `cloudService` field specified
+        .orElse(x.as[model.RuntimeConfigRequest.DataprocConfig]: Either[DecodingFailure, RuntimeConfigRequest]) //try decode as DataprocConfig
+        .orElse(x.as[model.RuntimeConfigRequest.GceConfig]) //try decode as GceConfig
+  }
+
+  implicit val clusterRequestDecoder: Decoder[ClusterRequest] = Decoder.instance {
+    c =>
+      for {
+        labels <- c.downField("labels").as[Option[Map[String, String]]]
+        jupyterExtensionUri <- c.downField("jupyterExtensionUri").as[Option[GcsPath]]
+        jupyterUserScriptUri <- c.downField("jupyterUserScriptUri").as[Option[UserScriptPath]]
+        jupyterStartUserScriptUri <- c.downField("jupyterStartUserScriptUri").as[Option[UserScriptPath]]
+        machineConfig <- c.downField("machineConfig").as[Option[model.RuntimeConfigRequest.DataprocConfig]]
```

Comment:
Just so I understand: this is currently assuming `DataprocConfig` so it's not using the logic in `runtimeConfigDecoder` yet, right?

Reply:
yep...